### PR TITLE
ipatests: Test for ipa-ca-install on replica

### PR DESCRIPTION
**Pull request by jayesh-garg**
Test on replica for ipa-ca-install with options
--no-host-dns,--skip-schema-check

Signed-off-by: Jayesh <jgarg@redhat.com>

### CODE REVIEW

File: ipatests/pytest_ipa/integration/tasks.py [hunk 1]
```diff
@@ -1462,19 +1462,21 @@ def install_kra(host, domain_level=None, first_instance=False, raiseonerr=True):
...
 def install_ca(host, domain_level=None, first_instance=False,
-               external_ca=False, cert_files=None, raiseonerr=True):
+               external_ca=False, cert_files=None, raiseonerr=True,
+               extra_args=()):
     if domain_level is None:
         domain_level = domainlevel(host)
     check_domain_level(domain_level)
     command = ["ipa-ca-install", "-U", "-p", host.config.dirman_password,
                "-P", 'admin', "-w", host.config.admin_password]
+    command.append(extra_args)
```
Comment (frasertweedale) [inline ipatests/pytest_ipa/integration/tasks.py]: This is wrong.
```
>>> t = ()
>>> command = ['ls']
>>> command.append(t)
>>> command
['ls', ()]
```

I think you want `command.extend(extra_args)` instead, a la:
```
>>> command = ['ls']
>>> command.extend(())
>>> command
['ls']
```

The call sites where `extra_args=...` is used must changed also, to a list or tuple.
(See comments below).  So before `command.extend(extra_args)`, ensure the type is correct:
```python
if not isinstance(extra_args, (tuple, list)):
    raise TypeError("extra_args must be tuple or list")
```

File: ipatests/pytest_ipa/integration/tasks.py [hunk 2]
```diff
@@ -1462,19 +1462,21 @@ def install_kra(host, domain_level=None, first_instance=False, raiseonerr=True):
...
     # First step of ipa-ca-install --external-ca
     if external_ca:
         command.append('--external-ca')
     # Continue with ipa-ca-install --external-ca
     if cert_files:
         for fname in cert_files:
-            command.extend(['--external-cert-file', fname])
+            command.append(['--external-cert-file', fname])
```
Comment (frasertweedale) [inline ipatests/pytest_ipa/integration/tasks.py]: `append` is wrong; `extend` was correct.

File: ipatests/test_integration/test_installation.py [hunk 3]
```diff
@@ -226,6 +226,23 @@ def test_replica2_ipa_kra_install(self):
...
+
+    num_replicas = 2
+
+    @classmethod
+    def install(cls, mh):
+        tasks.install_master(cls.master, setup_dns=False)
+
+    def test_replica_ca_install_with_no_host_dns(self):
+        tasks.install_replica(self.master, self.replicas[0], setup_ca=False)
+        tasks.install_ca(self.replicas[0], extra_args=("--no-host-dns"))
```
Comment (frasertweedale) [inline ipatests/test_integration/test_installation.py]: extra_args should be tuple or list, i.e.:
```python
extra_args=("--no-host-dns",)  # note trailing comma, this is what makes it a 1-tuple.

# OR 

extra_args=["--no-host-dns"]
```

File: ipatests/test_integration/test_installation.py [hunk 4]
```diff
@@ -226,6 +226,23 @@ def test_replica2_ipa_kra_install(self):
...
+
+    def test_replica_ca_install_with_skip_schema_check(self):
+        tasks.install_replica(self.master, self.replicas[1], setup_ca=False)
+        tasks.install_ca(self.replicas[1], extra_args=("--skip-schema-check"))
```
Comment (frasertweedale) [inline ipatests/test_integration/test_installation.py]: Here too.

**Comment by frasertweedale** [comment]
Removing need-review label and adding WIP.  Please put needs-review back when the PR has been updated with the requested changes.

File: ipatests/pytest_ipa/integration/tasks.py [hunk 5]
```diff
@@ -1461,13 +1461,16 @@ def install_kra(host, domain_level=None, first_instance=False, raiseonerr=True):
...
+def install_ca(
+        host, domain_level=None, first_instance=False, external_ca=False,
+        cert_files=None, raiseonerr=True, extra_args=()
+):
     if domain_level is None:
         domain_level = domainlevel(host)
     check_domain_level(domain_level)
     command = ["ipa-ca-install", "-U", "-p", host.config.dirman_password,
                "-P", 'admin', "-w", host.config.admin_password]
+    command.extend(extra_args)
```
Comment (frasertweedale) [inline ipatests/pytest_ipa/integration/tasks.py]: Can you please add a type check for `extra_args` before using it?

```python
if not isinstance(extra_args, (tuple, list)):
    raise TypeError("extra_args must be tuple or list")
```

File: ipatests/test_integration/test_installation.py [hunk 6]
```diff
@@ -227,6 +227,23 @@ def test_replica2_ipa_kra_install(self):
         super(TestInstallWithCA2, self).test_replica2_ipa_kra_install()
 
 
+class TestInstallCA(IntegrationTest):
+
```
Comment (flo-renaud) [inline ipatests/test_integration/test_installation.py]: Can you add comments to the code? For instance for the class
```
""" Tests for CA installation on a replica
"""
```
and for each test method
```
"""
Test for ipa-ca-install --no-host-dns on a replica
"""
```
and
```
"""
Test for ipa-ca-install --skip-schema-check on a replica
"""
```

**Comment by flo-renaud** [comment]
Hi @jayesh-garg 
The `.freeipa-pr-ci.yaml` file is still pointing to temp_commit.yaml, please link `ipatests/prci_definitions/gating.yaml` instead.

**Comment by flo-renaud** [comment]
Hi @jayesh-garg 
thanks for the commit, LGTM.

**Comment by flo-renaud** [comment]
master:

* ad3bf5042d35b2d16646170b09489fb964fbdb93 Test for ipa-ca-install on replica
* fb3c2c140220b7d6d88558f9bb0a485f47079a5e Nightly definations commit

**Comment by flo-renaud** [comment]
@jayesh-garg 
manual backports are required for the other branches.

**Comment by flo-renaud** [comment]
Backports:
#4090 on ipa-4-6
#4091 on ipa-4-7
#4092 on ipa-4-8

**Pull request closed without merging**